File: packages/runtime/src/metrics.ts
/**
 * The counters §16 asks to expose, in Prometheus text format on an internal
 * /metrics endpoint. No client library: a counter map and a serialiser are
 * eighty lines, and the alternative is a dependency that pulls in a registry, a
 * default-metrics collector and a cluster aggregator for a box with one user.
 */

type Labels = Record<string, string | number>;

interface Series {
  help: string;
  type: 'counter' | 'gauge' | 'histogram';
  values: Map<string, number>;
  buckets?: number[];
  sums?: Map<string, number>;
  counts?: Map<string, number>;
}

const registry = new Map<string, Series>();

function key(labels: Labels = {}): string {
  const entries = Object.entries(labels).sort(([a], [b]) => a.localeCompare(b));
  return entries.map(([k, v]) => `${k}="${String(v).replace(/"/g, '\\"')}"`).join(',');
}

function series(name: string, help: string, type: Series['type'], buckets?: number[]): Series {
  let s = registry.get(name);
  if (!s) {
    s = { help, type, values: new Map(), buckets, sums: new Map(), counts: new Map() };
    registry.set(name, s);
  }
  return s;
}

export function counter(name: string, help: string) {
  const s = series(name, help, 'counter');
  return {
    inc(labels: Labels = {}, by = 1): void {
      const k = key(labels);
      s.values.set(k, (s.values.get(k) ?? 0) + by);
    },
  };
}

export function gauge(name: string, help: string) {
  const s = series(name, help, 'gauge');
  return {
    set(value: number, labels: Labels = {}): void {
      s.values.set(key(labels), value);
    },
  };
}

export function histogram(name: string, help: string, buckets: number[]) {
  const s = series(name, help, 'histogram', buckets);
  return {
    observe(value: number, labels: Labels = {}): void {
      const base = key(labels);
      for (const b of buckets) {
        if (value <= b) {
          const k = `${base}${base ? ',' : ''}le="${b}"`;
          s.values.set(k, (s.values.get(k) ?? 0) + 1);
        }
      }
      const kInf = `${base}${base ? ',' : ''}le="+Inf"`;
      s.values.set(kInf, (s.values.get(kInf) ?? 0) + 1);
      s.sums!.set(base, (s.sums!.get(base) ?? 0) + value);
      s.counts!.set(base, (s.counts!.get(base) ?? 0) + 1);
    },
  };
}

export function renderMetrics(): string {
  const out: string[] = [];
  for (const [name, s] of registry) {
    out.push(`# HELP ${name} ${s.help}`);
    out.push(`# TYPE ${name} ${s.type}`);
    for (const [labels, value] of s.values) {
      const suffix = s.type === 'histogram' ? '_bucket' : '';
      out.push(`${name}${suffix}${labels ? `{${labels}}` : ''} ${value}`);
    }
    if (s.type === 'histogram') {
      for (const [labels, sum] of s.sums!) out.push(`${name}_sum${labels ? `{${labels}}` : ''} ${sum}`);
      for (const [labels, n] of s.counts!) out.push(`${name}_count${labels ? `{${labels}}` : ''} ${n}`);
    }
  }
  return `${out.join('\n')}\n`;
}

// The catalogue from §16, declared once so a typo in a label name is a compile
// error somewhere rather than a metric that silently never appears.
export const M = {
  messagesRead: counter('messages_read_total', 'Messages fetched from a provider'),
  messagesDropped: counter('messages_dropped_total', 'Messages the classifier dropped'),
  extraction: counter('extraction_total', 'Extraction attempts by rung and outcome'),
  modelLatency: histogram('model_latency_seconds', 'Rung 3 call latency', [0.5, 1, 2, 5, 10, 20, 30, 60]),
  modelFailures: counter('model_failures_total', 'Rung 3 failures by kind'),
  resolverDecisions: counter('resolver_decisions_total', 'Resolver outcomes'),
  reviewItemsOpen: gauge('review_items_open', 'Open review items'),
  denylistViolations: counter('denylist_violations_total', 'Article 9 fields dropped from model output'),
  eventsAppended: counter('events_appended_total', 'Events appended by type'),
  notificationsSent: counter('notifications_sent_total', 'Push notifications by rule'),
  mailboxFreshnessSeconds: gauge('mailbox_freshness_seconds', 'Seconds since a mailbox last read successfully'),
  queueDepth: gauge('queue_depth', 'Messages waiting per queue'),
  deadLetters: gauge('dead_letters', 'Messages in a dead-letter queue'),
};

File: backend/loop/api/routes/metrics.py
"""Four numbers, in the one format everything already scrapes.

Hand-rolled, and the reason is worth stating: a Prometheus client library brings
a registry, a multiprocess mode, a WSGI app and a set of default collectors, to
serialise a dozen lines of text. The exposition format is a documented text
format and this is a text formatter.

**Four, not the thirteen §16 lists, and the gap is deliberate for now.** The
four here — queue depth, dead letters, open review items, mailbox freshness —
are facts about Postgres, so they are read from Postgres on request: simpler
than aggregating counters across eight processes, and correct after a restart,
which an in-process counter is not.

The other nine are per-event counters that only the worker that emits the event
can see: `messages_read_total`, `messages_dropped_total`, `extraction_total`,
`model_latency_seconds`, `model_failures_total`, `resolver_decisions_total`,
`denylist_violations_total`, `events_appended_total`, `notifications_sent_total`.
The reference declared all nine and incremented them in the workers, and served
them from a per-worker health port that this port does not open — so restoring
them means giving the queue services an HTTP endpoint again, or writing the
counts somewhere Postgres can be asked about them. Until one of those happens
they are absent, and saying so here is better than a docstring that makes an
absence look like a scrape failure.
"""

from typing import Any, Final

from fastapi import APIRouter, Request
from fastapi.responses import PlainTextResponse

from loop.db import Queue, dead_letter_depth

router = APIRouter()

# Prometheus text exposition, version 0.0.4. No charset: the format's own media
# type carries the encoding.
CONTENT_TYPE: Final = "text/plain; version=0.0.4"

_SECONDS_PER_HOUR: Final = 3600


@router.get("/metrics", response_class=PlainTextResponse)
async def metrics(request: Request) -> PlainTextResponse:
    """Public, like every other scrape endpoint, and it says nothing personal.

    Counts and ages, no ids, no addresses, no company names. That is what makes
    leaving it open defensible rather than an oversight — but it is still a
    statement about how much mail this mailbox gets, so it belongs behind the
    same network boundary as the database.
    """
    lines: list[str] = []
    async with request.app.state.db.untenanted() as connection:
        depths = {}
        for queue in Queue.ALL:
            row = await connection.fetchrow("select * from mq.metrics($1)", queue)
            depths[queue] = int(row["queue_length"] or 0) if row else 0
        dead = await dead_letter_depth(connection)
        review = await connection.fetchval(
            "select count(*) from review_items where resolved_at is null"
        )
        freshness = await connection.fetchval(
            "select extract(epoch from now() - max(last_ok_at)) from mailbox_accounts"
        )

    _series(
        lines,
        "queue_depth",
        "gauge",
        "Messages waiting per queue",
        [({"queue": queue}, depth) for queue, depth in depths.items()],
    )
    _series(lines, "dead_letters", "gauge", "Messages in a dead-letter queue", [({}, dead)])
    _series(lines, "review_items_open", "gauge", "Open review items", [({}, int(review or 0))])
    if freshness is not None:
        _series(
            lines,
            "mailbox_freshness_seconds",
            "gauge",
            "Seconds since a mailbox last read successfully",
            [({}, float(freshness))],
        )
    # The content type is set as a header rather than through `media_type`,
    # which appends a charset to anything under `text/` — and Prometheus's
    # media type carries its version parameter there instead.
    return PlainTextResponse(
        "\n".join(lines) + "\n", headers={"content-type": CONTENT_TYPE}
    )


def _series(
    lines: list[str],
    name: str,
    kind: str,
    help_text: str,
    samples: list[tuple[dict[str, str], Any]],
) -> None:
    lines.append(f"# HELP {name} {help_text}")
    lines.append(f"# TYPE {name} {kind}")
    for labels, value in samples:
        lines.append(f"{name}{_labels(labels)} {value}")


def _labels(labels: dict[str, str]) -> str:
    if not labels:
        return ""
    pairs = ",".join(f'{key}="{_escaped(labels[key])}"' for key in sorted(labels))
    return f"{{{pairs}}}"


def _escaped(value: str) -> str:
    return value.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")
